pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,250 @@
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		crmetrics.Registry.MustRegister(gauge)
+	}
+}
+
+func (ns *nodeScraper) update(ctx context.Context) {
+	nodes := make(map[string]struct{})
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			logging.FromContext(ctx).Debugf("Tracking new node: %s", n.Node.Name)
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes[n.Node.Name] = struct{}{}
+
+		podRequests := resources.Subtract(n.PodTotalRequests, n.DaemonSetRequested)
+		podLimits := resources.Subtract(n.PodTotalLimits, n.DaemonSetLimits)
+		allocatable := n.Node.Status.Capacity
+		if len(n.Node.Status.Allocatable) > 0 {
+			allocatable = n.Node.Status.Allocatable
+		}
+
+		// Populate  metrics
+		var err error
+		for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+			overheadGaugeVec:       ns.getSystemOverhead(n.Node),
+			podRequestsGaugeVec:    podRequests,
+			podLimitsGaugeVec:      podLimits,
+			daemonRequestsGaugeVec: n.DaemonSetRequested,
+			daemonLimitsGaugeVec:   n.DaemonSetLimits,
+			allocatableGaugeVec:    allocatable,
+		} {
+			err = multierr.Append(err, ns.set(resourceList, n.Node, gaugeVec))
+		}
+
+		if err != nil {
+			logging.FromContext(ctx).Errorf("Failed to generate gauges for %s: %s", n.Node.Name, err)
+		}
+
+		return true
+	})
+
+	ns.cleanup(ctx, nodes)
+}
+
+func (ns *nodeScraper) reset() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		gauge.Reset()
+	}
+}
+
+func (ns *nodeScraper) cleanup(ctx context.Context, existingNodes map[string]struct{}) {
+	nodesToRemove := []string{}
+	for nodeName := range ns.labelMap {
+		if _, ok := existingNodes[nodeName]; !ok {
+			nodesToRemove = append(nodesToRemove, nodeName)
+		}
+	}
+
+	// Remove all gauges associated with removed node
+	for _, node := range nodesToRemove {
+		gaugeMap := ns.labelMap[node]
+		for gaugeVec, labelSet := range gaugeMap {
+			for _, labels := range labelSet {
+				gaugeVec.Delete(labels)
+			}
+		}
+		delete(ns.labelMap, node)
+	}
+
+	if len(nodesToRemove) > 0 {
+		logging.FromContext(ctx).Debugf("Removed the following node gauges: %s", strings.Join(nodesToRemove, ", "))
+	}
+}
+
+// set sets the value for the node gauge
+func (ns *nodeScraper) set(resourceList v1.ResourceList, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+	for resourceName, quantity := range resourceList {
+		// Reformat resource type to be consistent with Prometheus naming conventions (snake_case)
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+
+		labels := ns.getNodeLabels(node, resourceTypeName)
+		ns.labelMap[node.Name][gaugeVec] = append(ns.labelMap[node.Name][gaugeVec], labels)
+
+		gauge, err := gaugeVec.GetMetricWith(labels)
```

Comment:
There is a non-errorable form, which will clean up a lot of your error handling. 
e.g.
`podGaugeVec.With(labels).Set(float64(1))`

Reply:
Do we want the metrics code to panic with an error though? I don't think a metric error should be considered non-recoverable.

Reply:
IIUC this will panic early, often, and loudly, since this code will execute in all cases. I agree that you don't want edge cases to panic.

This prometheus library is more flexible than our use case. IIUC, this will only panic due to programmer error, and there is no user input or transient failures that could cause this to fail. Because of this, we can treat the fact that labels are mapped correct as an invariant in the code base.

This is similar to indexing a map when you know there will be a value, even if technically the map supports not having a value. Happy to chat over voice w/ these nuances.